### ISSUE TITLE
Add automatic derivation of `Imm` to `Deriver`

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#4848]
Add support for automatic derivation of `Imm` for immutable types. 

@mlutze Perhaps you can expand on what is needed. I also don't know whether to treat this as `Boxed` or more like `Eq`?

**Comment by mlutze** [comment]
This is a bit trickier than I had originally thought, since `Imm` is a marker interface, so there is no instance definition that we create to check all the enum components for free.

**Comment by magnus-madsen** [comment]
> This is a bit trickier than I had originally thought, since `Imm` is a marker interface, so there is no instance definition that we create to check all the enum components for free.

What does this mean?

**Comment by mlutze** [comment]
> > This is a bit trickier than I had originally thought, since `Imm` is a marker interface, so there is no instance definition that we create to check all the enum components for free.
> 
> What does this mean?

When we derive `Eq`, we generate an instance like:

```
enum Blah[a] with Eq {
    case Blah(a -> a)
}

// generated instance
instance Eq[Blah[a]] with Eq[a] {
    pub def eq(x: Blah[a], y: Blah[a]): Bool = match (x, y) {
        case (Blah(f), Blah(g)) => f == g                // ERROR: no instance of Eq for functions
    }
}
```
We get that error for free in the Typer.

With `Imm` the generated instance is like this:

```
enum Blah[a] with Imm {
    case Blah(a -> a)
}

instance Imm[Blah[a] with Imm[a]  // just a marker class so no traversal done
```

No free errors.

One hack I can think of is to introduce a useless signature that traverses the object:

```
class Imm[a] {
    pub def check(x: a): Unit
}
```

Otherwise we can have a more ad-hoc solution.

**Comment by paulbutcher** [comment]
How about this (which avoids having to create a meaningless `check` function in `Immutable`). We define a function `isImmutable`:

```
pub def isImmutable(_: a): Bool with Immutable[a] = true
```

and then for:

```
enum Foo[a](Bar, Qux, a) with Immutable
```

we generate:

```
instance Immutable[Foo[a]] with Immutable[a]

def _dummy1(x: Bar): Bool = isImmutable(x)
def _dummy2(x: Qux): Bool = isImmutable(x)
```

**Comment by paulbutcher** [comment]
Although, looking at the implementation of the deriver, this may require too many changes to be worth it (there's no obvious way to have the deriver generate code outside of instances). So perhaps the `check` member of `Immutable` is the lesser of the available evils.

**Comment by magnus-madsen** [comment]
I think this is moving a bit too fast: Why do we need all this complexity? I would expect a simple check.

If that is not readily possible, what about a hack where every type is immutable followed by a simple check in `Safety`?

In particular, the idea of generating functions (i.e. NEW CODE) is very surprising to me. 
We don't want to generate useless code, right?

**Comment by paulbutcher** [comment]
I completely agree, this is awful (I didn't say that in the PR because I didn't want to skew your impressions before you had a chance to look at it).

But, if we continue down this road then complexity of one kind or another is, I think, inevitable, because `Immutable` is abusing the concept of a type class. We're using it as a "marker", not as a way of implementing functionality and as @mlutze pointed out above, the current implementation of type classes isn't written with that in mind. We're trying to use it in a way which is more like an annotation than a type class.

There is nowhere as things currently stand to put a "simple" check. It can't be in the deriver because the deriver doesn't have all the information necessary to carry out the check: we have to check at some point later in the pipeline when we know exactly how our enum will be used, and there is no such current point. If we create such a point, we now have a choice about whether we use it only to handle `Immutable` (so `Immutable` is a special case) or whether we modify all derived implementations to work in the same way.

Bottom line: we're using a type class for something it's really not designed to do.

And worse, we're investing a lot of effort into something which doesn't even give us any value: as already discussed what we really need for the problem we're trying to solve is something closer to `ThreadSafe` than `Immutable`. I really don't see how `Immutable` gets us anywhere close to our ultimate goal.

**Comment by mlutze** [comment]
> But, if we continue down this road then complexity of one kind or another is, I think, inevitable, because Immutable is abusing the concept of a type class. We're using it as a "marker", not as a way of implementing functionality and as @mlutze pointed out above, the current implementation of type classes isn't written with that in mind. We're trying to use it in a way which is more like an annotation than a type class.

I think we still want an `Immutable` type class: In my view there is indeed a class of types that are `Immutable`, even if there is no associated functionality.

> There is nowhere as things currently stand to put a "simple" check. It can't be in the deriver because the deriver doesn't have all the information necessary to carry out the check: we have to check at some point later in the pipeline when we know exactly how our enum will be used, and there is no such current point. If we create such a point, we now have a choice about whether we use it only to handle Immutable (so Immutable is a special case) or whether we modify all derived implementations to work in the same way.

I don't totally understand what you mean here. It seems to me that a simple recursive check over the types in the cases of an enum is sufficient, no? Per a brief discussion with Magnus, if any of the types in the enum is not `Immutable` then the enum is not `Immutable`.

> And worse, we're investing a lot of effort into something which doesn't even give us any value: as already discussed what we really need for the problem we're trying to solve is something closer to ThreadSafe than Immutable. I really don't see how Immutable gets us anywhere close to our ultimate goal.

Is `Immutable` not useful whatsoever? My understanding was that it was a useful stepping stone.

Proposal:

* We keep the `Immutable` type class, without any signatures. 
* We add a check to the `Safety` phase that checks all the instances of `Immutable` by ensuring all their constituent types are regionless. (Or maybe that they implement `Immutable`?)

**Comment by magnus-madsen** [comment]
A general comment for Java interop. We always have three choices when Java does not have sufficient information:
- Require the programmer to annotate (i.e. instruct the compiler) about what it needs to know.
- Make pessimistic assumptions. (Maybe with an escape hatch, e.g. unsafe cast, etc).
- Make optimistic assumption. (Its fine, don't worry).

I would assume "most" Java classes are not thread-safe nor immutable. This rules out (3). Thus we have (1) and (2). I would vote for (2) not sending Java objects over channels. But there could be some escape hatch. The escape hatch could be some kind of wrapper, some kind of unsafe send, some kind of annotation/cast, or other compiler trickery.

**Comment by magnus-madsen** [comment]
We could have:

```
pub def send(m: t, s: Sender[t]): Unit \ IO with Immutable[t]
pub def unsafeSend(m: t, s: Sender[t]): Unit \ IO
```

This would work for both Java objects and thread-safe objects (until we have a better mechanism for the latter).

**Comment by paulbutcher** [comment]
Fair enough on the Java interop side. My point, though, was that if we go down this road and add checks for `Immutable` within safety then we either:

1. Have `Immutable` as a special case, different from all other type classes, or
2. Bring other type classes in line with `Immutable`

If we choose option 2, that will fix some issues where code compiles that probably shouldn't, but complicates situations where the programmer deliberately ignores some elements of an enum (which is might be perfectly legitimate).

Plus...

For every other type class, the programmer has the option of full control over how that type class is implemented. But that won't be true for `Immutable`.

**Comment by mlutze** [comment]
I think there's some confusion about what precisely the Deriver generates.

Reproducing what you have here:
```scala
enum Foo(Int32) with ToString // Does not derive Eq

enum Bar[a](a) with Eq, ToString

def foo(): Bar[Foo] = Bar(Foo(42))
```

The generated instance for `Bar` is:
```scala
instance Eq[Bar[a]] with Eq[a] { ... }
```
which means "Equality is defined on `Bar[a]` _only if_ it is defined on `a`.

This definition:
```scala
def foo(): Bar[Foo] = Bar(Foo(42))
```
is fine because `Eq` is not relevant. 

We can consider why we do it this way by considering `List`. We probably want to be able to make a `List[Foo]` even though there's no implementation of `Eq` for `Foo`.

**Comment by paulbutcher** [comment]
Ah! So by extension:

```
enum Bar[a](a) with Immutable
```

Should also mean `Bar` is immutable only if `a` is immutable? But then throwing an error in `Safety` would be the wrong thing to do? Instead it should silently remove the instance of `Immutable` for `Bar` with this particular subtype?

But I'm not really sure how that would work?

**Comment by paulbutcher** [comment]
Ah! I think I'm getting it. That falls out in the wash already. So the only types we would be checking during safety would be the concrete types used within the definition of the enum, not the type parameters?

**Comment by mlutze** [comment]
> Ah! I think I'm getting it. That falls out in the wash already. So the only types we would be checking during safety would be the concrete types used within the definition of the enum, not the type parameters?

Yes, that's right. The type parameters will be covered by the constraints on the instance. The other types are checked in `Safety`.

**Comment by mlutze** [comment]
`TypedAst.Instance` has a `tpe` field. We can check when this type looks like `MyEnum[a, b, c]`, then lookup the enum from there.
Specifically, `Type.baseType` is probably what you want here.

From there I think you should be able to use the `ClassEnvironment` object: Using the `classEnv` from the `TypedAst.Root`, you can verify that `Immutable[t]` holds for every `t` in the enum, using [`ClassEnvironment.holds`](https://github.com/flix/flix/blob/master/main/src/ca/uwaterloo/flix/language/phase/unification/ClassEnvironment.scala#L62).
